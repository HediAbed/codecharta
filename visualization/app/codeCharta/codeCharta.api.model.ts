import {
	AttributeTypes,
	AttributeTypeValue,
	CodeMapNode,
	Edge,
	MarkedPackage,
	MetricStatisticsItem,
	RecursivePartial,
	Settings
} from "./codeCharta.model"

export interface ExportCCFile {
	projectName: string
	apiVersion: string
	fileChecksum: string
	nodes: CodeMapNode[]
	attributeTypes?: AttributeTypes | OldAttributeTypes
	edges?: Edge[]
	markedPackages?: MarkedPackage[]
	blacklist?: ExportBlacklistItem[]
	metricStatistics?: ExportMetricStatisticsItem[]
}

export interface ExportBlacklistItem {
	path: string
	type: ExportBlacklistType
}

export type ExportMetricStatisticsItem = MetricStatisticsItem

export enum ExportBlacklistType {
	hide = "hide",
	exclude = "exclude"
}

export enum APIVersions {
	ZERO_POINT_ONE = "0.1",
	ONE_POINT_ZERO = "1.0",
	ONE_POINT_ONE = "1.1",
	ONE_POINT_TWO = "1.2"
}

export interface ExportScenario {
	name: string
	settings: RecursivePartial<Settings>
}

export interface OldAttributeTypes {
	nodes?: [{ [key: string]: AttributeTypeValue }?]
	edges?: [{ [key: string]: AttributeTypeValue }?]
}
