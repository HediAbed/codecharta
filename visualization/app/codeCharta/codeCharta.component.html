<tool-bar-component></tool-bar-component>

<file-extension-bar-component ></file-extension-bar-component>
<div style="margin-top:1px; width: 100%; min-height: 100px; position: fixed; z-index: 10; overflow: hidden; white-space: nowrap;font-size: 0px;" >	
	<ribbon-bar-component class="ribbon-bar-component"></ribbon-bar-component >
</div>


<div id="actionButtons">
	<md-button
		ng-if="$ctrl._viewModel.focusedNodePath.length > 0"
		class="md-fab md-primary dark_shadow_opacity long-button"
		aria-label="Remove Focus on node"
		ng-click="$ctrl.removeFocusedNode()"
	>
		Show complete map
	</md-button>
	<md-button class="md-fab md-primary dark_shadow_opacity" aria-label="Fit Map to View" ng-click="$ctrl.fitMapToView()">
		<i class="fa fa-compress fa-lg"></i>
	</md-button>
</div>

<code-map-component></code-map-component>

<legend-panel-component></legend-panel-component>

<detail-panel-component></detail-panel-component>

<loading-gif-file-component></loading-gif-file-component>

<div id="mw-logo">
	<div class="logo">
		<a href="https://www.maibornwolff.de" target="_blank"><img src="assets/logo.png" alt="MaibornWolff"/></a>
		<h2>
			<span>CodeCharta {{ $ctrl._viewModel.version }}</span>
		</h2>
	</div>
</div>

<node-context-menu-component></node-context-menu-component>
