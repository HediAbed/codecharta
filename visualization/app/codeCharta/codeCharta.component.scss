@import "../mixins.scss";

code-charta-component {
	$animationDuration: 0.5;

	#mw-logo {
		.logo {
			width: 160px;
			height: 100px;
			position: absolute;
			left: 0;
			bottom: 30px;
			z-index: 29;
		}

		a,
		img {
			width: 5em;
			display: block;
			margin: auto !important;
		}

		h2 {
			text-transform: uppercase;
			color: #5a585a;
			font-family: Roboto, sans-serif;
			font-weight: 500;
			font-size: 9pt;
			text-align: center;
			margin: 0;
			z-index: -1;
		}
	}

	#title {
		position: fixed;
		z-index: 2;
		width: 100%;
	}

	.small-info {
		color: rgba(0, 0, 0, 0.38);
		font-size: 9pt;
	}

	#codeMap {
		position: fixed;
		z-index: 0;
		top: 90px;
		left: 0;
		width: 100%;
		height: 100%;
		transition: all $animationDuration + s ease;

		&.expanded {
			height: calc(100% - 170px);
		}
	}

	#actionButtons {
		position: fixed;
		bottom: 2em;
		right: 2em;
		z-index: 60;

		i.fa-folder-open {
			margin: 4px 0 0 2px;
		}

		.long-button {
			width: inherit;
			padding-left: 10px;
			padding-right: 10px;
			border-radius: 100px;
		}

		> * {
			margin-left: 4px;
			margin-right: 4px;
			z-index: 60;
		}
	}

	.dark_shadow_opacity {
		box-shadow: 0 1px 5px 0 rgba(0, 0, 0, 0.2) !important;
		opacity: 0.97;
	}

	.bottom_shadow {
		box-shadow: 0 4px 4px -2px rgba(0, 0, 0, 0.2) !important;
	}

	.md-button.round-button {
		margin-right: 0 !important;
		float: right;
		min-width: 36px;
		width: 36px;
		border-radius: 50%;
	}

	.ribbon-bar-component {
		z-index: 10;

		md-card,
		.section .section-header,
		.section .section-body,
		#toggle-ribbon-bar-fab {
			transition: all $animationDuration + s ease;
		}
	}

	.ribbon-bar-wrapper {
		margin-top: 1px;
		width: 100%;
		min-height: 100px;
		height: 500px;
		position: fixed;
		z-index: 10;
		overflow: hidden;
		white-space: nowrap;
		font-size: 0px;
		pointer-events: none;
	}

	.angular-green {
		color: #00bfa5;
	}

	.small-action-button {
		height: 24px;
		width: 24px;
		float: left;

		*,
		.md-fab:active {
			margin: 0;
			padding: 0;
		}

		.md-fab {
			position: relative;
			min-height: 0;
			width: 100%;
			height: 100%;
			z-index: 0;
			border-radius: 50%;

			i {
				position: absolute;
				top: 5px;
				left: 6px;
				color: white;

				&:hover {
					cursor: pointer;
					color: #eee;
				}
			}
		}
	}
}
