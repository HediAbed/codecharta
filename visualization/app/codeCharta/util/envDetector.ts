export function isStandalone(): boolean {
	return JSON.parse(process.env.STANDALONE)
}
