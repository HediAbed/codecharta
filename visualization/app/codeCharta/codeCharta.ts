"use strict";

import "./core/core.module";
import "./ui/ui";

import {codeChartaComponent} from "./codeCharta.component";

import angular from "angular";
import { CodeChartaService } from "./codeCharta.service";
import _ from "lodash";

angular.module(
    "app.codeCharta",
    ["app.codeCharta.core", "app.codeCharta.ui"]
);

angular.module("app.codeCharta").component(
    codeChartaComponent.selector,
    codeChartaComponent
).service(
    _.camelCase(CodeChartaService.name),
    CodeChartaService
);


