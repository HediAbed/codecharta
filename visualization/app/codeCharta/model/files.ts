import { CCFile, FileSelectionState, FileState } from "./codeCharta.model"

export class Files {
	private files: FileState[] = []

	public getFiles(): FileState[] {
		return this.files
	}

	public reset() {
		this.files = []
	}

	public resetSelection() {
		this.files.map(file => {
			file.selectedAs = FileSelectionState.None
		})
	}

	public getCCFiles(): CCFile[] {
		return this.files.map(x => x.file)
	}

	public addFile(file: CCFile) {
		this.files.push({ file: file, selectedAs: FileSelectionState.None })
	}

	public setSingleByName(fileName: string) {
		const singleFile: CCFile = this.getFileByFileName(fileName)
		this.setSingle(singleFile)
	}

	public setSingle(file: CCFile) {
		this.resetSelection()
		const matchedFile = this.files.find(x => x.file === file)
		if (matchedFile) {
			matchedFile.selectedAs = FileSelectionState.Single
		}
	}

	public setDeltaByNames(referenceFileName: string, comparisonFileName: string) {
		const referenceFile: CCFile = this.getFileByFileName(referenceFileName)
		const comparisonFile: CCFile = this.getFileByFileName(comparisonFileName)
		this.setDelta(referenceFile, comparisonFile)
	}

	public setDelta(reference: CCFile, comparison: CCFile) {
		this.resetSelection()
		const matchedReferenceFile = this.files.find(x => x.file === reference)
		const matchedComparisonFile = this.files.find(x => x.file === comparison)

		if (matchedReferenceFile) {
			matchedReferenceFile.selectedAs = FileSelectionState.Reference
		}

		if (matchedComparisonFile) {
			matchedComparisonFile.selectedAs = FileSelectionState.Comparison
		}
	}

	public setMultipleByNames(partialFileNames: string[]) {
		const partialFiles: CCFile[] = []
		partialFileNames.forEach(fileName => partialFiles.push(this.getFileByFileName(fileName)))
		this.setMultiple(partialFiles)
	}

	public setMultiple(multipleFiles: CCFile[]) {
		this.resetSelection()
		this.files.filter(x => multipleFiles.indexOf(x.file) !== -1).forEach(x => (x.selectedAs = FileSelectionState.Partial))
	}

	public fileStatesAvailable(): boolean {
		return this.files.length > 0
	}

	public getFileByFileName(fileName: string): CCFile {
		const matchingFileState: FileState = this.files.find(x => x.file.fileMeta.fileName == fileName)
		return matchingFileState ? matchingFileState.file : undefined
	}

	public getVisibleFiles(): CCFile[] {
		return this.files.filter(x => x.selectedAs !== FileSelectionState.None).map(x => x.file)
	}

	public getVisibleFileStates(): FileState[] {
		return this.files.filter(x => x.selectedAs !== FileSelectionState.None)
	}

	public isSingleState(): boolean {
		return !!this.files.find(x => x.selectedAs === FileSelectionState.Single)
	}

	public isDeltaState(): boolean {
		return !!this.files.find(x => x.selectedAs === FileSelectionState.Reference || x.selectedAs === FileSelectionState.Comparison)
	}

	public isPartialState(): boolean {
		return !!this.files.find(x => x.selectedAs === FileSelectionState.Partial)
	}
}
