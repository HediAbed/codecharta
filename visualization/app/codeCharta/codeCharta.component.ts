import { NameDataPair, UrlUtils } from "./util/urlUtils"
import {IHttpService, ILocationService, IRootScopeService} from "angular"
import "./codeCharta.component.scss"
import { NodeContextMenuController } from "./ui/nodeContextMenu/nodeContextMenu.component"
import { CodeMapActionsService } from "./ui/codeMap/codeMap.actions.service"
import { CodeChartaService } from "./codeCharta.service"

/**
 * This is the main controller of the CodeCharta application
 */
export class CodeChartaController {
	public viewModel = {
		version: require("../../package.json").version,
		numberOfLoadingTasks: 0
	}
	private urlUtils: UrlUtils

	/* @ngInject */
	constructor(
		//private threeOrbitControlsService: ThreeOrbitControlsService,
		private $rootScope: IRootScopeService,
		//private dialogService: DialogService,
		//private codeMapActionsService: CodeMapActionsService,
		private codeChartaService: CodeChartaService,
		private $location: ILocationService,
		private $http: IHttpService
	) {
		this.urlUtils = new UrlUtils(this.$location, this.$http)
		this.subscribeToLoadingEvents(this.$rootScope)
		this.loadFileOrSample()
	}

	public fitMapToView() {
		// this.threeOrbitControlsService.autoFitTo()
	}

	public removeFocusedNode() {
		//this.codeMapActionsService.removeFocusedNode()
	}

	public loadFileOrSample() {
		this.viewModel.numberOfLoadingTasks++
		return this.urlUtils.getFileDataFromQueryParam().then((data: NameDataPair[]) => {
			if (data.length > 0) {
				this.tryLoadingFiles(data)
			} else {
				this.tryLoadingSampleFiles()
			}
		}, this.tryLoadingSampleFiles.bind(this))
	}

	public tryLoadingSampleFiles() {
		if (this.urlUtils.getParam("file")) {
			/*this.dialogService.showErrorDialog(
				"One or more files from the given file URL parameter could not be loaded. Loading sample files instead."
			)*/
		}
		this.tryLoadingFiles([
            { name: "sample1.json", data: require("./assets/sample1.json") },
            { name: "sample2.json", data: require("./assets/sample2.json") }
        ]);
    }
    
    private tryLoadingFiles(values: NameDataPair[]) {
        this.codeChartaService
        .loadFiles(
            values
        )
        .then(() => this.viewModel.numberOfLoadingTasks--)
        .catch(e => {
			this.viewModel.numberOfLoadingTasks--
			console.error(e);
            this.printErrors(e)
        })
    }

	public printErrors(errors: Object) {
		//this.dialogService.showErrorDialog(JSON.stringify(errors, null, "\t"))
	}

	private subscribeToLoadingEvents($rootScope: angular.IRootScopeService) {
		$rootScope.$on("add-loading-task", () => {
			this.viewModel.numberOfLoadingTasks++
		})

		$rootScope.$on("remove-loading-task", () => {
			this.viewModel.numberOfLoadingTasks--
		})
	}
}

export const codeChartaComponent = {
	selector: "codeChartaComponent",
	template: require("./codeCharta.component.html"),
	controller: CodeChartaController
}
