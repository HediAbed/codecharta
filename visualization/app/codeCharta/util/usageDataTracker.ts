"use strict"

import { getVisibleFileStates, isSingleState } from "../model/files/files.helper"
import { FileExtensionCalculator, MetricDistribution } from "./fileExtensionCalculator"
import { CodeChartaStorage } from "./codeChartaStorage"
import { MetricKeyValue, MetricsPerLanguage, State } from "../codeCharta.model"
import { isStandalone } from "./envDetector"
import { isActionOfType } from "./reduxHelper"
import { AreaMetricActions } from "../state/store/dynamicSettings/areaMetric/areaMetric.actions"
import { HeightMetricActions } from "../state/store/dynamicSettings/heightMetric/heightMetric.actions"
import { ColorMetricActions } from "../state/store/dynamicSettings/colorMetric/colorMetric.actions"
import { ColorRangeActions } from "../state/store/dynamicSettings/colorRange/colorRange.actions"
import { BlacklistActions } from "../state/store/fileSettings/blacklist/blacklist.actions"
import { InvertColorRangeActions } from "../state/store/appSettings/invertColorRange/invertColorRange.actions"
import { FocusedNodePathActions } from "../state/store/dynamicSettings/focusedNodePath/focusedNodePath.actions"
import md5 from "md5"
import { getMedian } from "./nodeDecorator"

interface TrackingDataItem {
	mapId: string
	codeChartaApiVersion: string
	creationTime: number
	exportedFileSizeInBytes: number
	totalLinesOfCode: number
	totalRealLinesOfCode: number
	programmingLanguages: string[]
	totalNumberOfFiles: number
	numberOfFilesPerLanguage: Map<string, number>
	maxFilePathDepth: number
	avgFilePathDepth: number
	languageDistribution: LanguageDistributionStatistics
	metricStatisticsPerLanguage: MetricStatisticsPerLanguage
	metricStatisticsOverall: MetricStatisticsOverall
}

interface LanguageDistributionStatistics {
	[languageName: string]: {
		metricName: string
		absoluteValue: number
		relativeValue: number
	}
}

interface MetricStatisticsPerLanguage {
	[languageName: string]: {
		[metricName: string]: TrackingDataMetricItem
	}
}

interface MetricStatisticsOverall {
	[metricName: string]: TrackingDataMetricOutlierItem
}

interface TrackingDataMetricItem {
	metricName: string
	min: number
	max: number
	totalSum: number
	numberOfFiles: number
	median: number
	avg: number
}

interface TrackingDataMetricOutlierItem extends TrackingDataMetricItem {
	boxplot_irq_outliers: string[]
}

function isTrackingAllowed(state: State) {
	return isStandalone() && isSingleState(state.files) && getVisibleFileStates(state.files).length === 1
}

export function trackMetaUsageData(state: State) {
	if (!isTrackingAllowed(state)) {
		return
	}

	const singleFileStates = getVisibleFileStates(state.files)

	const fileMeta = singleFileStates[0].file.fileMeta
	const fileMetaStatistics = fileMeta.statistics

	// TODO: handle and check for absence of rloc and presence of loc
	const languageDistribution = FileExtensionCalculator.getMetricDistribution(singleFileStates[0].file.map, "rloc")

	const trackingDataItem: TrackingDataItem = {
		mapId: fileMeta.fileChecksum,
		codeChartaApiVersion: fileMeta.apiVersion,
		creationTime: Date.now(),
		exportedFileSizeInBytes: fileMeta.exportedFileSize,
		languageDistribution: mapLanguageDistribution(languageDistribution),
		totalRealLinesOfCode: fileMetaStatistics.totalRealLinesOfCode,
		totalLinesOfCode: fileMetaStatistics.totalLinesOfCode,
		programmingLanguages: fileMetaStatistics.programmingLanguages,
		numberOfFilesPerLanguage: fileMetaStatistics.numberOfFilesPerLanguage,
		totalNumberOfFiles: fileMetaStatistics.totalNumberOfFiles,
		maxFilePathDepth: fileMetaStatistics.maxFilePathDepth,
		avgFilePathDepth: fileMetaStatistics.avgFilePathDepth,
		metricStatisticsPerLanguage: mapMetricStatisticsPerLanguage(fileMetaStatistics.metricStatisticsPerLanguage),
		metricStatisticsOverall: mapMetricStatisticsOverall(fileMetaStatistics.metricStatisticsOverall)
	}

	const fileStorage = new CodeChartaStorage()

	try {
		fileStorage.setItem(`usageData/${trackingDataItem.mapId}-meta`, JSON.stringify(trackingDataItem))
	} catch {
		// ignore it
	}
}

function mapLanguageDistribution(languageDistribution: MetricDistribution[]) {
	const mappedDistribution: LanguageDistributionStatistics = {}

	for (const distributionItem of languageDistribution) {
		mappedDistribution[distributionItem.fileExtension] = {
			metricName: "rloc",
			absoluteValue: distributionItem.absoluteMetricValue,
			relativeValue: distributionItem.relativeMetricValue
		}
	}

	return mappedDistribution
}

function mapMetricStatisticsPerLanguage(metricStatisticsPerLanguage: MetricsPerLanguage) {
	//TODO: reuse mapMetricStatisticsOverall()
	const mappedStatistics = {}

	for (const languageIndex of Object.keys(metricStatisticsPerLanguage)) {
		const languageMetrics = {}
		mappedStatistics[languageIndex] = languageMetrics

		for (const metricIndex of Object.keys(metricStatisticsPerLanguage[languageIndex])) {
			const metricStatistics = metricStatisticsPerLanguage[languageIndex][metricIndex]

			// Filter 0-values to calculate median properly
			let metricValues = metricStatistics.values
			metricValues = metricValues.filter(function (metricValue) {
				return metricValue > 0
			})

			languageMetrics[metricIndex] = {
				metricName: metricIndex,
				min: metricStatistics.min,
				max: metricStatistics.max,
				totalSum: metricStatistics.totalSum,
				numberOfFiles: metricStatistics.numberOfFiles,
				median: getMedian(metricValues).toPrecision(2),
				avg: metricStatistics.average.toPrecision(2)
			}
		}
	}

	return mappedStatistics
}

function mapMetricStatisticsOverall(metricStatisticsOverall: MetricKeyValue) {
	const mappedStatistics = {}

	for (const metricIndex of Object.keys(metricStatisticsOverall)) {
		const metricStatistics = metricStatisticsOverall[metricIndex]

		// Filter 0-values to calculate median properly
		let metricValues = metricStatistics.values
		metricValues = metricValues.filter(function (metricValue) {
			return metricValue > 0
		})

		mappedStatistics[metricIndex] = {
			metricName: metricIndex,
			boxplot_irq_outliers: [],
			min: metricStatistics.min,
			max: metricStatistics.max,
			totalSum: metricStatistics.totalSum,
			numberOfFiles: metricStatistics.numberOfFiles,
			median: getMedian(metricValues).toPrecision(2),
			avg: metricStatistics.average.toPrecision(2)
		}
	}
	return mappedStatistics
}

interface SettingChangedEventPayload {
	eventName: string
	newValue: any
}

interface NodeInteractionEventPayload {
	eventName: string
	id: string
	type?: string
	nodeType?: string
	attributes?: MetricStatisticsOverall
}

interface EventTrackingItem {
	eventType: string
	eventTime: number
	payload: SettingChangedEventPayload | NodeInteractionEventPayload
}

export function trackEventUsageData(actionType: string, state: State, payload?: any) {
	if (!isTrackingAllowed(state)) {
		return
	}

	//TODO: Refactor
	if (
		!isActionOfType(actionType, AreaMetricActions) &&
		!isActionOfType(actionType, HeightMetricActions) &&
		!isActionOfType(actionType, ColorMetricActions) &&
		!isActionOfType(actionType, ColorRangeActions) &&
		!isActionOfType(actionType, InvertColorRangeActions) &&
		!isActionOfType(actionType, BlacklistActions) &&
		!isActionOfType(actionType, FocusedNodePathActions)
	) {
		return
	}

	let eventTrackingItem: EventTrackingItem

	if (
		isActionOfType(actionType, AreaMetricActions) ||
		isActionOfType(actionType, HeightMetricActions) ||
		isActionOfType(actionType, ColorMetricActions) ||
		isActionOfType(actionType, ColorRangeActions) ||
		isActionOfType(actionType, InvertColorRangeActions)
	) {
		eventTrackingItem = {
			eventType: "setting_changed",
			eventTime: Date.now(),
			payload: {
				eventName: actionType,
				newValue: payload
			}
		}
	}

	if (isActionOfType(actionType, BlacklistActions)) {
		eventTrackingItem = {
			eventType: "node_interaction",
			eventTime: Date.now(),
			payload: {
				eventName: actionType,
				id: md5(payload.path),
				type: payload.type,
				nodeType: payload.nodeType,
				attributes: payload.attributes
			}
		}
	}

	if (isActionOfType(actionType, FocusedNodePathActions) && payload !== "") {
		eventTrackingItem = {
			eventType: "node_interaction",
			eventTime: Date.now(),
			payload: {
				eventName: actionType,
				id: md5(payload)
			}
		}
	}

	const { fileChecksum } = getVisibleFileStates(state.files)[0].file.fileMeta
	const fileStorage = new CodeChartaStorage()

	try {
		const appendedEvents = `${fileStorage.getItem(`usageData/${fileChecksum}-events`)}\n${JSON.stringify(eventTrackingItem)}`
		fileStorage.setItem(`usageData/${fileChecksum}-events`, appendedEvents)
	} catch {
		// ignore tracking errors
	}
}
