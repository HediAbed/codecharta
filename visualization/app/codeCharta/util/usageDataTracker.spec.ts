import { CCFile, CodeMapNode, NodeType, State } from "../codeCharta.model"
import { trackMapMetaData } from "./usageDataTracker"
import * as EnvironmentDetector from "./envDetector"
import * as FilesHelper from "../model/files/files.helper"
import { FileState } from "../model/files/files"
import { APIVersions } from "../codeCharta.api.model"
import { CodeChartaStorage } from "./codeChartaStorage"
jest.mock("./codeChartaStorage")

describe("UsageDataTracker", () => {
	describe("trackMetaUsageData", () => {
		// provide some default state properties
		const stateStub = {
			appSettings: {
				experimentalFeaturesEnabled: true,
				showMetricLabelNameValue: undefined,
				isWhiteBackground: false,
				camera: { x: 1, y: 2, z: 3 }
			},
			dynamicSettings: {},
			fileSettings: {}
		} as State

		const setItemMock = jest.fn()
		CodeChartaStorage.prototype.setItem = setItemMock

		let singleFileState: FileState
		beforeEach(() => {
			jest.resetAllMocks()

			singleFileState = {
				file: {
					fileMeta: {
						fileChecksum: "invalid-md5-sample-checksum",
						apiVersion: APIVersions.ONE_POINT_ONE,
						exportedFileSize: 999
					},
					map: {}
				} as CCFile
			} as FileState
		})

		it("should not track in web version, in multi/delta mode or for more than one uploaded file", () => {
			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(false)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([{} as FileState])

			trackMapMetaData(stateStub)

			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(false)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([{} as FileState])

			trackMapMetaData(stateStub)

			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([{} as FileState, {} as FileState])

			trackMapMetaData(stateStub)

			expect(setItemMock).not.toHaveBeenCalled()
		})

		it("should not track maps from old API versions", () => {
			singleFileState.file.fileMeta.apiVersion = APIVersions.ONE_POINT_ZERO

			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])

			trackMapMetaData(stateStub)

			singleFileState.file.fileMeta.apiVersion = "0.2"
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])

			trackMapMetaData(stateStub)

			expect(setItemMock).not.toHaveBeenCalled()
		})

		it("should track from API version 1.1", () => {
			singleFileState.file.map = { path: "/root" } as CodeMapNode

			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])

			trackMapMetaData(stateStub)

			singleFileState.file.fileMeta.apiVersion = "1.2"
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])

			trackMapMetaData(stateStub)

			singleFileState.file.fileMeta.apiVersion = "2.0"
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])

			trackMapMetaData(stateStub)

			expect(setItemMock).toHaveBeenCalledTimes(3)
		})

		it("should track files with multiple programming languages properly", () => {
			singleFileState.file.map = mapStub

			jest.spyOn(EnvironmentDetector, "isStandalone").mockReturnValue(true)
			jest.spyOn(FilesHelper, "isSingleState").mockReturnValue(true)
			jest.spyOn(FilesHelper, "getVisibleFileStates").mockReturnValue([singleFileState])

			jest.spyOn(Date, "now").mockReturnValue(1612369999999)

			const expectSetItemSnapshot = (CodeChartaStorage.prototype.setItem = jest.fn().mockImplementation((_, value) => {
				expect(value).toMatchSnapshot()
			}))

			trackMapMetaData(stateStub)

			expect(expectSetItemSnapshot).toHaveBeenCalledTimes(1)
		})
	})

	const mapStub: CodeMapNode = {
		name: "root",
		attributes: {},
		type: NodeType.FOLDER,
		isExcluded: false,
		isFlattened: false,
		children: [
			{
				name: "doNotTrackInvalidFileWithMissingExtension",
				path: "/root/doNotTrackInvalidFileWithMissingExtension",
				type: NodeType.FILE,
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				isExcluded: false,
				isFlattened: false,
				link: "http://www.google.de"
			},
			{
				name: "bigLeaf.ts",
				path: "/root/bigLeaf.ts",
				type: NodeType.FILE,
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				isExcluded: false,
				isFlattened: false,
				link: "http://www.google.de"
			},
			{
				name: "bigLeaf.java",
				path: "/root/bigLeaf.java",
				type: NodeType.FILE,
				attributes: { rloc: 100, functions: 10, mcc: 1 },
				isExcluded: false,
				isFlattened: false,
				link: "http://www.google.de"
			},
			{
				name: "Parent Leaf",
				path: "/root/Parent Leaf",
				type: NodeType.FOLDER,
				attributes: {},
				isExcluded: false,
				isFlattened: false,
				children: [
					{
						name: "smallLeaf.ts",
						path: "/root/Parent Leaf/smallLeaf.ts",
						type: NodeType.FILE,
						attributes: { rloc: 30, functions: 100, mcc: 100 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "otherLeaf.ts",
						path: "/root/Parent Leaf/otherLeaf.ts",
						type: NodeType.FILE,
						attributes: { rloc: 70, functions: 1000, mcc: 10 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "smallLeaf.java",
						path: "/root/Parent Leaf/smallLeaf.java",
						type: NodeType.FILE,
						attributes: { rloc: 30, functions: 100, mcc: 100 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "otherLeaf.java",
						path: "/root/Parent Leaf/otherLeaf.java",
						type: NodeType.FILE,
						attributes: { rloc: 70, functions: 1000, mcc: 10 },
						isExcluded: false,
						isFlattened: false
					},
					{
						name: "anotherLeaf.java",
						path: "/root/Parent Leaf/anotherLeaf.java",
						type: NodeType.FILE,
						attributes: { rloc: 20, functions: 10, mcc: 2 },
						isExcluded: false,
						isFlattened: false
					}
				]
			}
		]
	}
})
