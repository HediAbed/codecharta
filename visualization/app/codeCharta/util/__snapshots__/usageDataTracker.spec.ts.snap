// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsageDataTracker trackMetaUsageData should track files with multiple programming languages properly 1`] = `"{\\"mapId\\":\\"invalid-md5-sample-checksum\\",\\"codeChartaApiVersion\\":\\"1.1\\",\\"creationTime\\":1612369999999,\\"exportedFileSizeInBytes\\":999,\\"statisticsPerLanguage\\":{\\"ts\\":{\\"metrics\\":{\\"rloc\\":{\\"avg\\":66.66666666666667,\\"max\\":100,\\"median\\":70,\\"min\\":30,\\"numberOfFiles\\":3,\\"totalSum\\":200},\\"functions\\":{\\"avg\\":370,\\"max\\":1000,\\"median\\":100,\\"min\\":10,\\"numberOfFiles\\":3,\\"totalSum\\":1110},\\"mcc\\":{\\"avg\\":37,\\"max\\":100,\\"median\\":10,\\"min\\":1,\\"numberOfFiles\\":3,\\"totalSum\\":111}},\\"numberOfFiles\\":3,\\"maxFilePathDepth\\":4,\\"avgFilePathDepth\\":3.6666666666666665},\\"java\\":{\\"metrics\\":{\\"rloc\\":{\\"avg\\":55,\\"max\\":100,\\"median\\":50,\\"min\\":20,\\"numberOfFiles\\":4,\\"totalSum\\":220},\\"functions\\":{\\"avg\\":280,\\"max\\":1000,\\"median\\":55,\\"min\\":10,\\"numberOfFiles\\":4,\\"totalSum\\":1120},\\"mcc\\":{\\"avg\\":28.25,\\"max\\":100,\\"median\\":6,\\"min\\":1,\\"numberOfFiles\\":4,\\"totalSum\\":113}},\\"numberOfFiles\\":4,\\"maxFilePathDepth\\":4,\\"avgFilePathDepth\\":3.75}}}"`;
